Sources/SwiftOperatorPrecedence/PrecedenceGraph.swift

```diff
@@ -0,0 +1,149 @@
+    // the graph. This detects cycles and prevents extraneous work.
+    var groupsSeen: Set<PrecedenceGroupName> = []
+
+    var stack: [(PrecedenceGroupName, SyntaxProtocol?)] =
+      [(initialGroupName, initialSyntax)]
+    while let (currentGroupName, currentGroupSyntax) = stack.popLast() {
+      guard let currentGroup = lookupGroup(currentGroupName) else {
+        try errorHandler(
+          .missingGroup(currentGroupName, referencedFrom: currentGroupSyntax))
+        continue
+      }
+
+      for relation in currentGroup.relations {
+        if relation.kind == direction {
+          // If we hit our initial group, we're done.
+          let otherGroupName = relation.groupName
+          if otherGroupName == targetGroupName {
+            switch direction {
+            case .lowerThan:
+              return .lowerThan
+
+            case .higherThan:
+              return .higherThan
+            }
+          }
+
+          if groupsSeen.insert(otherGroupName).inserted {
+            stack.append((otherGroupName, relation.syntax))
+          }
+        }
+      }
+    }
+
+    return nil
+  }
+
+  /// Determine the precedence relationship between two precedence groups.
+  ///
+  /// Follow the precedence relationships among the precedence groups to
+  /// determine the precedence of the start group relative to the end group.
+  ///
+  /// - Returns: Precedence.lowerThan if startGroupName has lower precedence
+  /// than endGroupName, Precedence.higherThan if startGroupName has higher
+  /// precedence than endGroup name, and Precedence.unrelated otherwise.
+  func precedence(
+    relating startGroupName: PrecedenceGroupName,
+    to endGroupName: PrecedenceGroupName,
+    startSyntax: SyntaxProtocol?,
+    endSyntax: SyntaxProtocol?,
+    errorHandler: OperatorPrecedenceErrorHandler = { throw $0 }
+  ) rethrows -> Precedence {
+    if startGroupName == endGroupName {
+      return .unrelated
+    }
+
+    // Walk all of the relationships from the end down, then from the beginning
+    // up, to determine whether there is a relation between the two groups.
+    return try searchRelationships(
+      initialGroupName: endGroupName, initialSyntax: endSyntax,
+      targetGroupName: startGroupName, direction: .lowerThan,
+      errorHandler: errorHandler
+    ) ?? searchRelationships(
+      initialGroupName: startGroupName, initialSyntax: startSyntax,
+      targetGroupName: endGroupName, direction: .higherThan,
+      errorHandler: errorHandler
+    ) ?? searchRelationships(
+      initialGroupName: startGroupName, initialSyntax: startSyntax,
+      targetGroupName: endGroupName, direction: .lowerThan,
+      errorHandler: errorHandler
+    ).map {
+      $0.flipped
+    } ?? searchRelationships(
+      initialGroupName: endGroupName, initialSyntax: endSyntax,
+      targetGroupName: startGroupName, direction: .higherThan,
+      errorHandler: errorHandler
+    ).map {
+      $0.flipped
+    } ?? .unrelated
```

Comment:
Reading this I initially thought that `$0.flipped` matched all previous checks (my brain somehow thought that `??`) has higher precedence than `.`. Just wondering whether this would read easier as follows (I’m also not a huge fan of the 20 line expression)
```suggestion
    if let relationship = try searchRelationships(
      initialGroupName: endGroupName, initialSyntax: endSyntax,
      targetGroupName: startGroupName, direction: .lowerThan,
      errorHandler: errorHandler
    ) {
      return relationship
    } else if let relationship = try searchRelationships(
      initialGroupName: startGroupName, initialSyntax: startSyntax,
      targetGroupName: endGroupName, direction: .higherThan,
      errorHandler: errorHandler
    ) {
      return relationship
    } else if let relationship = try searchRelationships(
      initialGroupName: startGroupName, initialSyntax: startSyntax,
      targetGroupName: endGroupName, direction: .lowerThan,
      errorHandler: errorHandler
    ) {
      return relationship.flipped
    } else if relationship = try searchRelationships(
      initialGroupName: endGroupName, initialSyntax: endSyntax,
      targetGroupName: startGroupName, direction: .higherThan,
      errorHandler: errorHandler
    ) {
      return relationship.flipped
    } else {
      return .unrelated
    }
```

Reply:
To my eye, both look equally messy. I'd rather find a better answer, perhaps with a real data structure.